tests/test_json_schema.py

```diff
@@ -2378,15 +2389,28 @@ def __get_pydantic_core_schema__(
             cls, source: Any, handler: Callable[[Any], core_schema.CoreSchema]
         ) -> core_schema.CoreSchema:
             if hasattr(source, '__args__'):
-                param = source.__args__[0]
-                metadata = build_metadata_dict(js_cs_override=handler(Optional[param]))
+                arg = source.__args__[0]
+
+                def js_func(s, h):
+                    # ignore the schema we were given and get a new CoreSchema
+                    s = handler(Optional[arg])
+                    return h(s)
+
                 return core_schema.general_plain_validator_function(
                     Gen,
-                    metadata=metadata,
+                    metadata={'pydantic_js_functions': [js_func]},
```

Comment:
This is one area of the API that is still kludgy: `build_metadata_dict` used in this test is private; we shouldn't use private stuff in tests. Can we make `metadata` a `TypedDict` with `total=False` or something?

Reply:
I think metadata has the same problem as the ConfigDict — when I implemented this, I thought that using a typeddict directly meant that users/libraries/etc. wouldn't be able to add fields. But I think we realized we could solve that through having people create subclasses. So yeah, maybe that's a better way to do it, but I'll just note that I don't know how well it would work if trying to use two packages together that have custom metadata (maybe you can do diamond inheritance?)

At any rate, I'm in favor of eliminating the build_metadata_dict and using a typeddict, and just changing that API in the future if we need to.